### PR TITLE
Handle "strict" in behat configuration file

### PR DESCRIPTION
**Pull request by jdeniau**
Follows https://github.com/Behat/Behat/pull/1501 and #1512 

Implement the fact that we can set the `strict` settings in the behat.yml in `config:` section.

### CODE REVIEW

File: src/Behat/Behat/Config/Handler/StrictHandler.php
Comment (acoulton) [inline src/Behat/Behat/Config/Handler/StrictHandler.php:29]: Now that we only support PHP ^8.1, we should probably use typed properties / constructor promotion / readonly properties in new code? (we can't yet use readonly classes as they are 8.2 only).

```suggestion

    public function __construct(
        private readonly ResultInterpreter $resultInterpreter,
    ) {}
```

File: src/Behat/Testwork/Tester/Cli/StrictController.php
Comment (acoulton) [inline src/Behat/Testwork/Tester/Cli/StrictController.php:43]: As above, I'd suggest using strict types for new code (but leave the existing props with `@var` declarations for now - modernising old code is a separate concern IMO).

```suggestion
    private StrictHandler $strictHandler;
```

**Review by acoulton — COMMENTED** [review]
LGTM, just a couple of suggestions on using new PHP features now that we can, but maybe wait to see if @carlos-granados and @ttomdewit agree with me before applying them as this would be the start of a new convention.

**Comment by acoulton** [comment]
It would be great if you could also submit a PR to add the stop on failure & strict config options to https://github.com/Behat/docs/blob/v3.0/user_guide/configuration.rst? Thanks

**Comment by jdeniau** [comment]
> It would be great if you could also submit a PR to add the stop on failure & strict config options to [Behat/docs@v3.0/user_guide/configuration.rst](https://github.com/Behat/docs/blob/v3.0/user_guide/configuration.rst?rgh-link-date=2024-11-04T08%3A14%3A50Z)? Thanks

@acoulton yes, it's on my mind, I just wanted to add `strict` mode first to avoid making two different PR on the documentation 👍

File: src/Behat/Behat/Config/Handler/StrictHandler.php
Comment (jdeniau) [reply src/Behat/Behat/Config/Handler/StrictHandler.php:29]: I was wondering what was the policy on that kind of feature.

If you did drop older version of PHP, that's great, and I can of course migrate typing. You might also consider migrate all codebase to typed property using rectorphp and this kind of set: https://getrector.com/find-rule?rectorSet=core-type-declarations&activeRectorSetGroup=core (but if you consider that it might be breaking, it may be only on the next major).

Another "new" php feature I was wondering is the `array('foo')` vs `['foo']`. Behat does use `array` everywhere, but I think that the standard is more `['foo`]` now (but I kept `array` to be consistent too).

**Comment by jdeniau** [comment]
@acoulton I did change to more modern PHP code. I converted the the StopOnFailure handler from #1501 as it has not been released in a version for now.

File: src/Behat/Behat/Config/Handler/StrictHandler.php
Comment (carlos-granados) [inline src/Behat/Behat/Config/Handler/StrictHandler.php:17]: Should be something like "Enables strict via config"

**Review by carlos-granados — CHANGES_REQUESTED** [review]
Just a small change needed, otherwise looking good

**Review by carlos-granados — APPROVED** [review]
Looking good now, thanks @jdeniau

**Comment by jdeniau** [comment]
> It would be great if you could also submit a PR to add the stop on failure & strict config options to [Behat/docs@`v3.0`/user_guide/configuration.rst](https://github.com/Behat/docs/blob/v3.0/user_guide/configuration.rst?rgh-link-date=2024-11-04T08%3A14%3A50Z)? Thanks

https://github.com/Behat/docs/pull/155

**Comment by carlos-granados** [comment]
@jdeniau I took a new look at your PR and thought of something: instead of handling the strict flag and config setting in the stop on failure controller and handler, why not just pass the result interpreter to the handler and use it to check the interpretation? This would provide a better separation of concerns as all the handling of the strict flag and config setting would be just in the strict controller and handler. What do you think?

**Comment by jdeniau** [comment]
@carlos-granados Well I am not really confident with `ResultInterpreter` because it does register multiple interpretations, and because it does ["Interprets result as a UNIX return code (0 for success, 1 for failure)."](https://github.com/jdeniau/Behat/blob/be2b1432969dce440ba67f7a7aa451f01a05292c/src/Behat/Testwork/Tester/Result/ResultInterpreter.php#L38) : it seems less readable to check `=== 0` instead of `=== true`.

I will try an implementation in another PR 🤞

**Comment by jdeniau** [comment]
@carlos-granados You were right, this seems to be a way better implementation: https://github.com/Behat/Behat/pull/1516

**Comment by ttomdewit** [comment]
Thanks for the contribution!

Please do not wait on my review and/or approval, I haven't been able to dig deep enough to fully understand what is going on. I trust @acoulton @carlos-granados when they approve this.

**Comment by acoulton** [comment]
Closed in favour of #1516

**Pull request closed without merging**
